monitoring/opencensus/main.go

```diff
@@ -0,0 +1,125 @@
+// Sets up metrics.
+var (
+	requestCount       = stats.Int64("oc_request_count", "total request count", "requests")
+	failedRequestCount = stats.Int64("oc_failed_request_count", "count of failed requests", "requests")
+	responseLatency    = stats.Float64("oc_latency_distribution", "distribution of response latencies", "s")
+)
+
+// [END monitoring_sli_metrics_opencensus_measure]
+// [START monitoring_sli_metrics_opencensus_view]
+// Sets up views.
+var (
+	requestCountView = &view.View{
+		Name:        "oc_request_count",
+		Measure:     requestCount,
+		Description: "total request count",
+		Aggregation: view.Count(),
+	}
+	failedRequestCountView = &view.View{
+		Name:        "oc_failed_request_count",
+		Measure:     failedRequestCount,
+		Description: "count of failed requests",
+		Aggregation: view.Count(),
+	}
+	responseLatencyView = &view.View{
+		Name:        "oc_response_latency",
+		Measure:     responseLatency,
+		Description: "The distribution of the latencies",
+		// Bucket definitions must be explicitly specified.
+		Aggregation: view.Distribution(0, 1000, 2000, 3000, 4000, 5000, 6000, 7000, 8000, 9000, 10000),
+	}
+)
+
+// [END monitoring_sli_metrics_opencensus_view]
+
+func main() {
+
+	// Expects that the project ID be provided via a flag when starting the server.
+	projectID := flag.String("project_id", "", "Cloud Project ID")
+	flag.Parse()
+	// [START monitoring_sli_metrics_opencensus_view]
+	// Register the views.
+	if err := view.Register(requestCountView, failedRequestCountView, responseLatencyView); err != nil {
+		log.Fatalf("Failed to register the views: %v", err)
+	}
+	// [END monitoring_sli_metrics_opencensus_view]
+	// [START monitoring_sli_metrics_opencensus_exporter]
+	// Sets up Cloud Monitoring exporter.
+	sd, err := stackdriver.NewExporter(stackdriver.Options{
+		ProjectID:         *projectID,
+		MetricPrefix:      "opencensus-demo",
+		ReportingInterval: 60 * time.Second,
+	})
+	if err != nil {
+		log.Fatalf("Failed to create the Cloud Monitoring exporter: %v", err)
+	}
+	defer sd.Flush()
+
+	sd.StartMetricsExporter()
+	defer sd.StopMetricsExporter()
+	// [END monitoring_sli_metrics_opencensus_exporter]
+	http.HandleFunc("/", handle)
+	log.Fatal(http.ListenAndServe(":8080", nil))
+}
+
+func handle(w http.ResponseWriter, r *http.Request) {
+	ctx, _ := tag.New(context.Background())
+	requestReceived := time.Now()
+	// Records latency for failure OR success.
+	// [START monitoring_sli_metrics_opencensus_latency]
+	defer func() {
+		stats.Record(ctx, responseLatency.M(time.Since(requestReceived).Seconds()))
+	}()
+	// [END monitoring_sli_metrics_opencensus_latency]
+	// [START monitoring_sli_metrics_opencensus_counts]
+	// Counts the request.
+	stats.Record(ctx, requestCount.M(1))
+
+	// Randomly fails 10% of the time.
+	if rand.Intn(100) > 90 {
+		// Counts the error.
+		stats.Record(ctx, failedRequestCount.M(1))
+		// [END monitoring_sli_metrics_opencensus_counts]
+		fmt.Fprintf(w, "intentional error!")
```

Comment:
That's how I had it originally.  However, that then requires that the test expects a 500 some of the time, which could obfuscate, for example, breaking changes, since it would just treat crashes as expected.  So, I wrote the test to validate the body response and return a 200 in either case.

Reply:
I suppose it doesn't really matter because it isn't part of the sample. But, the only way there would be a 500 is if it gets set in your handler. Any errors or panics or anything would show up in a different way. So, you should be confident that's the only place the 500 comes from.